6259 TOGGLE DAPI Stegvelger manglende innbetaling

fredrik-oh opened the pull request:
https://jira.adeo.no/browse/MELOSYS-6259
La til delvis støtte for saksbehandling av manglende innbetaling bak toggle. Steg 1 implementert, vedtakssteg ved opphør av medlemskap ikke implementert. Justerte labeltekst i periodesteget.

Enkelstegvelgers oppdaterStatus metode er endret til å støtte overstyring av flyt gjennom alleSteg array. Et steg kan nå sende med en stegId som spesifiserer neste steg som skal legges til i aktuelleSteg. Sendes stegId til oppdaterStatus fjernes alle steg etter steget som oppdaterer fra aktuelleSteg.

Commit:
6259 La til ny toggle for saksbehandling av manglende innbetaling og justerte lagUrl logikk

janitamarielle commented on src/url/url.ts:
```diff
@@ -123,14 +131,21 @@ export const skalViseIngenFlyt = (
-  if ([HENVENDELSE, KLAGE, MANGLENDE_INNBETALING_TRYGDEAVGIFT].includes(behandlingstype)) {
-    return true;
+  if (manglendeInnbetalingToggleEnabled) {
+    if ([HENVENDELSE, KLAGE].includes(behandlingstype)) {
```
Jeg hadde kanskje antatt det kom sjekk på type MANGLENDE_INNBETALING_TRYGDEAVGIFT her istedenfor.

janitamarielle replied:
Kanskje en if (manglendeInnbetalingToggleEnabled && behandlingstype == MANGLENDE_INNBETALING)

Må sikkert ikke det med en gang, men vil det være nyttig etterhvert å ha egen funksjon på dette i url?

Commit:
La på ny toggle parameter på alle lagUrl kall

Commit:
Merge branch 'master' into MELOSYS-6259_Stegvelger_manglende_innbetaling

Commit:
6259 TOGGLE DAPI Stegvelger manglende innbetaling

La til delvis støtte for saksbehandling av manglende innbetaling bak toggle. Steg 1 implementert, vedtakssteg ved opphør av medlemskap ikke implementert. Justerte labeltekst i periodesteget.

Enkelstegvelgers oppdaterStatus metode er endret til å støtte overstyring av flyt gjennom alleSteg array. Et steg kan nå sende med en stegId som spesifiserer neste steg som skal legges til i aktuelleSteg. Sendes stegId til oppdaterStatus fjernes alle steg etter steget som oppdaterer fra aktuelleSteg.

janitamarielle commented on src/felleskomponenter/enkelStegvelger/enkelStegvelger.tsx:
```diff
@@ -33,25 +33,31 @@ export default ({ alleSteg }: EnkelStegvelgerProps) => {
-  const oppdaterStatus = (stegId: string) => (isSchemaValid: boolean) => {
-    const nyeSteg = aktuelleSteg?.map((steg: AktueltSteg) =>
+  // Optional param nyttStegId overstyrer vanlig flyt.
+  const oppdaterStatus = (stegId: string) => (isSchemaValid: boolean, nyttStegId?: string) => {
```
Har du vurdert å bare ha egen liste over steg dersom behandlingstypen er MANGLENDE_INNBETALING? Altså den alleSteg-listen som sendes inn i enkeltStegvelger?

janitamarielle replied:
Ah, ser du har det. Hvorfor trengte man nyttStegId her da? Sliter litt med å se det 👀

fredrik-oh replied:
Commit kommentaren min som endret oppdaterStatus forklarer hvordan nyttStegId funker. 
For MANGLENDE_INNBETALING har vi to steg 1 i stegmappet som sendes inn til enkelStegvelger og avhengig av hva som fylles ut i inngangssteget sendes neste stegs id i oppdaterStatus kallet som kjøres.

janitamarielle replied:
Vurdert å ha alleSteg som er avhengig av både behandlingstype og fullstendigManglendeInnbetaling?

janitamarielle replied:
Bare litt bekymret for hva det betyr å ha flere steg underliggende som ikke vises, men som er der

fredrik-oh replied:
Jepp, det var tanken jeg startet med. Blir veldig kronglete mtp. function currying. OppdaterStatus funksjonen som sendes ned til stegene bindes til den initielle stegmappen som sendes inn til enkelStegvelger. Kan sikkert utdype litt muntlig hvis ønskelig

fredrik-oh replied:
oppdaterStatus er eneste kallet som legger til flere steg i aktuelleSteg. Kan evt. legge på en bool sjekk som sier at hvis du kaller oppdaterStatus på et vedtakssteg legges det ikke til et nytt steg

janitamarielle replied:
Kan oppdaterStatus da heller sende inn den nye alleSteg?

fredrik-oh replied:
Det er mulig ja, men ser det som en mer inngripende løsning. Gjør vi den optional på samme måte som nyttStegId er det effektivt samme løsning, men da har stegkomponentene enda mer interaksjon med implementasjonsdetaljer (hele stegmappet vs id på de neste relevante stegene) som burde være utenfor deres kontroll. Dersom vi gjør det mandatory å sende med stegmappet må vi jo endre alle konsumenter av enkelStegvelger som er enda mindre ønskelig.

janitamarielle replied:
Jeg syns ikke den skal bli mandatory nei. 

Det er bare veldig vanskelig å lese det nå, siden stegvelger er designet for enkle flyter, mens vi nå begynner med tre-struktur på flytene. 
Tror det derfor er lettere å lese "Skift til denne flyten []" istedenfor "Dette er neste steg, den finnes i listen fra før av". Selv med kommentarer er det ikke veldig lett å lese at saksbehandling/initialStegArrayManglendeInnbetaling er et tre (kombinasjon av to flyter) istedenfor en liste (én flyt).

fredrik-oh replied:
Tok diskusjonen muntlig. Ble enige om ekstra kommentar i oppdaterStatus metoden og rydding i stegarrayaet for å belyse at det er to mulige flyter gjennom stegene

janitamarielle commented on src/sider/ftrl/saksbehandling/stegKomponenter/vurderingVedtakOpphoer.tsx:
```diff
@@ -0,0 +1,19 @@
+  return (
+    <div className="vurderingVedtak">
+      <Nav.Typo.Innholdstittel className="stegvelgertittel">
+        Opphøer av frivillig medlemskap etter § 2-15
```
Opphør*?

janitamarielle commented on src/sider/ftrl/saksbehandling/stegKomponenter/vurderingInngangManglendeInnbetaling.tsx:
```diff
@@ -0,0 +1,103 @@
+  } = useForm({
+    resolver: yupResolver(vurdering_inngang_manglende_innbetaling),
+    mode: "all",
+    values: useMemo(() => initialValues as FieldValues, [initialValues]),
```
Trengte vi useMemo på denne?

fredrik-oh replied:
Nope, fikser.

Commit:
Vedtakssteg for fullstendig manglende innbetaling. Mapping av brevet OPPHOERT_MEDLEMSKAP mangler i api så dette kan anses som wip kode.

Commit:
Fjernet en skrivefeil og unødvendig bruk av useMemo i inngangsstegets useForm kall

janitamarielle commented on src/sider/ftrl/saksbehandling/initialStegArrayManglendeInnbetaling.tsx:
```diff
@@ -0,0 +1,100 @@
+  komponent: VurderingInngang,
+};
+
+const virksomhetSteg = {
```
Trengs disse duplikeres her, eller blir det mer lesbart om disse importeres fra den andre stegArrayen?

fredrik-oh replied:
Da må vi i så fall endre stegPosisjonen på stegene som importeres

janitamarielle replied:
Skjønner

janitamarielle commented on src/sider/ftrl/saksbehandling/initialStegArrayManglendeInnbetaling.tsx:
```diff
@@ -0,0 +1,100 @@
+  komponent: VurderingTrygdeavgift,
+};
+
+const vedtakSteg = {
```
Skal man ha både vedtakSteg og vedtakOpphoerSteg?
Edit: Dette er ikke lett å lese. Kan det hende vi vil dele disse opp i to? alleStegManglendeInnbetaling er jo ikke en flyt. Men heller en samling av to flyter

fredrik-oh replied:
Skal ha vedtakSteg og vedtakOpphoerSteg ja

janitamarielle replied:
Dette også avhengig av hva man velger på steg 1?

fredrik-oh replied:
Jepp, ser du på den oppdaterte koden har jeg lagt ved kommentarer og justert på rekkefølgen av stegene i filen så det forhåpentligvis er litt mer lesbart

janitamarielle replied:
Hehe, nesten så vi må rename "enkelStegvelger"

fredrik-oh replied:
:cry:

janitamarielle commented on src/sider/ftrl/saksbehandling/stegKomponenter/vurderingInngangManglendeInnbetaling.tsx:
```diff
@@ -0,0 +1,103 @@
+
+      <div className="label__container">
+        <Nav.Typo.Normaltekst>
+          Vurder konsekvens av manglende innbetaling. Du må sjekke OEBS for å se om betaling er mottatt innen fristen.{" "}
```
Funky {" "}

fredrik-oh replied:
Oj, intellij har gjort noe gøy

Commit:
Justerte på rekkefølgen i initialStegArrayManglendeInnbetaling.tsx og la til kommentarer som forklarer de to flytene

Commit:
Fjernet noe IDE rusk

Commit:
Justerte innholdet i initialStegArrayManglendeInnbetaling.tsx for å belyse at det er en trestruktur. La til en kommentar i oppdaterStatus som forklarer invalidering av etterfølgende steg
